kernel/src/drivers/ps2/io.rs

```diff
@@ -1,105 +1,27 @@
+pub static DATA_PORT: IOPort = unsafe { IOPort::new(0x60) };
+pub static STATUS_PORT: IOPort = unsafe { IOPort::new(0x64) };
+pub static COMMAND_PORT: IOPort = unsafe { IOPort::new(0x64) };
 
 pub const WAIT_TIMEOUT: u16 = 1000;
 
-const OUTPUT_STATUS_BIT: u8 = 1 << 0;
-const INPUT_STATUS_BIT: u8 = 1 << 1;
-
-/// Sends a controller command without a return
-pub fn command(cmd: ControllerCommand) {
-    command_raw(cmd as u8);
-}
-
-/// Sends a controller command with data and without a return
-pub fn command_data(cmd: ControllerCommand, data: u8) {
-    command_raw(cmd as u8);
-    write(&DATA_PORT, data);
-}
-
-/// Sends a controller command with a return
-pub fn command_ret(cmd: ControllerReturnCommand) -> Option<u8> {
-    command_raw(cmd as u8);
-    read(&DATA_PORT)
-}
-
-/// Sends a raw controller command code
-fn command_raw(cmd: u8) {
-    write(&COMMAND_PORT, cmd);
-}
-
-/// Writes the given value to the data port
-pub fn write_data(value: u8) {
-    write(&DATA_PORT, value);
-}
-
-/// Reads from the data port
-pub fn read_data() -> Option<u8> {
-    read(&DATA_PORT)
-}
-
-/// Writes to the given port, or waits until available
-pub fn write(port: &IOPort, value: u8) {
-    wait_write();
-    port.write(value);
-}
-
-/// Reads from the given port, returning an optional value. None returned if nothing could be read
-pub fn read(port: &IOPort) -> Option<u8> {
-    if wait_read() {
-        Some(port.read())
-    } else {
-        None
-    }
-}
-
-/// Flushes the controller's output buffer
-pub fn flush_output() {
-    loop {
-        // Read until the output status bit is empty
-        if check_status(OUTPUT_STATUS_BIT) {
-            DATA_PORT.read();
-        } else {
-            break;
-        }
-    }
-}
-
-/// Waits for the write status bit to empty
-fn wait_write() {
-    loop {
-        // Check if the input status bit is empty
-        if !check_status(INPUT_STATUS_BIT) {
-            break;
-        }
+bitflags! {
+    pub struct StatusFlags: u8 {
+        /// If the output buffer from the controller is full (data can be read)
+        const OUTPUT_FULL = 1 << 0;
+        /// If the input buffer to the controller is full (data cannot be written)
+        const INPUT_FULL = 1 << 1;
+        /// If the current output from the controller is from the second port
+        const OUTPUT_PORT_2 = 1 << 5;
     }
 }
 
-/// Waits for the read status bit to equal 1, and returns true if successful
-fn wait_read() -> bool {
-    for _i in 0..WAIT_TIMEOUT {
-        // Check if the output status bit is full
-        if check_status(OUTPUT_STATUS_BIT) {
-            return true;
-        }
-    }
-    false
-}
-
-/// Returns true if the given status bit is 1
-fn check_status(bit: u8) -> bool {
-    (STATUS_PORT.read() & bit) != 0
-}
-
 /// Represents a PS2 controller command without a return value
-#[allow(dead_code)]
+#[allow(dead_code)] // Dead variants for completeness
 #[derive(Copy, Clone)]
 #[repr(u8)]
 pub enum ControllerCommand {
```

Comment:
Maybe these command enum a could be moved into another mod (mod block keyword), such as ControllerCommand, or just Command. IMO since there are no other command types (correct me if I'm wrong) we could drop controller from the front
Also,  how come there isn't an enable port 1 variant?

Reply:
Ah, ok. Still, they could be moved into modules (inline, with mod keyword)

Reply:
Is there a reason for this? I don't believe we use this anywhere else.

Reply:
It helps to group them together in a module of their own. AFAICT we haven't had reason to do this up until now.

Reply:
What I mean is, why not group them into a separate file as well?

Reply:
You could if you wanted to, but there probably isn't enough code to justify moving into whole new file.